presto-spi/src/main/java/io/prestosql/spi/tracer/DefaultTracer.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.spi.tracer;
+
+import io.prestosql.spi.eventlistener.TracerEvent;
+
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+
+// Tracer keeps track of Presto execution unit contexts and sends debugging purpose events via its emitter.
+// Users should not build complex metrics based on the action events.
```

Comment:
What do you mean by "Users should not build complex metrics based on the action events." ?

Reply:
The tracer events should be used for developing debugging or diagnostic purposes only, and not be used to build metrics for production.